src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LanguageServerExportProviderBuilder.cs

```diff
@@ -0,0 +1,117 @@
+    {
+        var baseDirectory = AppContext.BaseDirectory;
+
+        // Load any Roslyn assemblies from the extension directory
+        using var _ = ArrayBuilder<string>.GetInstance(out var assemblyPathsBuilder);
+        assemblyPathsBuilder.AddRange(Directory.EnumerateFiles(baseDirectory, "Microsoft.CodeAnalysis*.dll"));
+        assemblyPathsBuilder.AddRange(Directory.EnumerateFiles(baseDirectory, "Microsoft.ServiceHub*.dll"));
+
+        // DevKit assemblies are not shipped in the main language server folder
+        // and not included in ExtensionAssemblyPaths (they get loaded into the default ALC).
+        // So manually add them to the MEF catalog here.
+        if (devKitDependencyPath != null)
+            assemblyPathsBuilder.AddRange(devKitDependencyPath);
+
+        // Add the extension assemblies to the MEF catalog.
+        assemblyPathsBuilder.AddRange(extensionManager.ExtensionAssemblyPaths);
+
+        // Create a MEF resolver that can resolve assemblies in the extension contexts.
+        var builder = new LanguageServerExportProviderBuilder(
+            assemblyPathsBuilder.ToImmutableAndClear(),
+            new Resolver(assemblyLoader),
+            cacheDirectory,
+            catalogPrefix: "c#-languageserver",
+            loggerFactory);
+        var exportProvider = await builder.CreateExportProviderAsync(cancellationToken);
+
+        // Also add the ExtensionAssemblyManager so it will be available for the rest of the composition.
+        exportProvider.GetExportedValue<ExtensionAssemblyManagerMefProvider>().SetMefExtensionAssemblyManager(extensionManager);
+
+        // Immediately set the logger factory, so that way it'll be available for the rest of the composition
+        exportProvider.GetExportedValue<ServerLoggerFactory>().SetFactory(loggerFactory);
+
+        return exportProvider;
+    }
+
+    protected override void LogError(string message)
+        => _logger.LogError(message);
+
+    protected override void LogTrace(string message)
+        => _logger.LogTrace(message);
+
+    protected override Task<ExportProvider> CreateExportProviderAsync(CancellationToken cancellationToken)
+    {
+        // Clear any previous cache write task, so that it is easy to discern whether
+        // a cache write was attempted.
+        s_cacheWriteTask = null;
+
+        return base.CreateExportProviderAsync(cancellationToken);
+    }
+
+    protected override bool ContainsUnexpectedErrors(IEnumerable<string> erroredParts, ImmutableList<PartDiscoveryException> partDiscoveryExceptions)
+    {
+        // Verify that we have exactly the MEF errors that we expect.  If we have less or more this needs to be updated to assert the expected behavior.
+        var expectedErrorPartsSet = new HashSet<string>(["CSharpMapCodeService", "PythiaSignatureHelpProvider", "CopilotSemanticSearchQueryExecutor"]);
+        var hasUnexpectedErroredParts = erroredParts.Any(part => !expectedErrorPartsSet.Contains(part));
```

Comment:
could literally be a switch, vs a set.  but this is ok.

Reply:
Seems a bit more concise in this form, but I don't feel strongly.